src/Features/Core/Portable/Diagnostics/AbstractHostDiagnosticUpdateSource.cs

```diff
@@ -77,8 +77,10 @@ public void ReportAnalyzerDiagnostic(DiagnosticAnalyzer analyzer, DiagnosticData
 
         public void ClearAnalyzerReferenceDiagnostics(AnalyzerFileReference analyzerReference, string language, ProjectId projectId)
         {
-            var analyzers = analyzerReference.GetAnalyzers(language);
-            ClearAnalyzerDiagnostics(analyzers, projectId);
+            if (analyzerReference.TryGetAnalyzers(language, out var analyzers))
```

Comment:
I think a simpler fix here would be to just check if the underlying assembly at `analyzerReference.FullPath` is loaded. You can then avoid all other changes in this PR.

Reply:
I'm not sure that would be sufficient to avoid the overhead of probing the assembly.

Reply:
The other option would be to keep a map of `AnalyzerFileReference`&rarr;`DiagnosticAnalyzer` somewhere within this type so we can avoid the `GetAnalyzers` call altogether.

Reply:
> we can avoid the GetAnalyzers call altogether.

This was the only call to `GetAnalyzers` so I'm not sure how that would help.

Reply:
I think we can implement this without public API changes. Given that we control and provide a [specialized analyzer assembly loader](http://sourceroslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/ProjectSystem/VisualStudioAnalyzer.cs,23) implementation to `AnalyzerFileReference`, we know when a specific analyzer assembly is requested to be loaded in the loader.

1. Update [ShadowCopyAnalyzerAssemblyLoader.LoadImpl](http://sourceroslyn.io/#Microsoft.CodeAnalysis.Workspaces/Diagnostics/CompilerShared/ShadowCopyAnalyzerAssemblyLoader.cs,6e66f40f99713804,references) override to track the `fullPath` that it gets invoked on.
2. Add an internal API `ShadowCopyAnalyzerAssemblyLoader.IsLoaded(string fullPath)`
3. Update `ClearAnalyzerReferenceDiagnostics` to take an additional argument `ShadowCopyAnalyzerAssemblyLoader analyzerAssemblyLoader`  and check `analyzerAssemblyLoader.IsLoaded(analyzerReference.FullPath)` here.

Alternatively, you can define a public event `public event EventHandler<AnalyzerLoadedEventArgs>? AnalyzerLoaded;`, so the subscription and tracking is done on the client side, say within `VisualStudioAnalyzer` itself, and it can guard the call to `ClearAnalyzerReferenceDiagnostics` with the is loaded check.

Reply:
Those solutions scare me a little because the `ShadowCopyAnalyzerAssemblyLoader` is necessarily static, so reasoning about what action a `VisualStudioAnalyzer` instance should take from it is difficult. If an old instance loaded an analyzer successfully, then a second instance could still load during cleanup.

I think I'm just going to close this out, as VS Mac has updated their side, and given #43008 cleaning this up doesn't make sense to me.